// models/index.js


/**
 * @module
 * This module facilitates the conversion between the different data models present in the application
 */
module.exports = {
    Banner: {

        
        classSearchToDB: {
            /**
             * Maps the properties in the Banner section object to the properties in the database model
             */
            sectionMap: {
                'crn': 'courseReferenceNumber',
                'title': 'courseTitle',
                'currentSeats': 'enrollment',
                'totalSeats': 'maximumEnrollment',
                'currentWaitlist': 'waitCount',
                'totalWaitlist': 'waitCapacity',
                'isOpen': 'openSection',
                'section': 'sequenceNumber'
            },

            classTimeMap: {
                
            },

            dayMap: {
                'monday': 'M',
                'tuesday': 'T',
                'wednesday': 'W',
                'thursday': 'R',
                'friday': 'F',
                'saturday': 'S',
                'sunday': 'U'
             },

            /**
             * Converts the data returned by Banner.ClassSearch() to the database model
             * @param {BannerResponse} response 
             */
            convert: function(response){
                let sections = response.data;

            },

            /**
             * Gets the name of the professor for the section
             * @param {BannerSection} section 
             */
            getInstructor: function(section){
                return section.faculty.length > 0 ?
                    section.faculty.find(f => f.primaryIndicator).displayName :
                    null;
            },

            /**
             * Extracts an array of days from the meeting time
             * @param {meetingTime} meetingTime 
             */
            getDays: function(meetingTime){
                return Object.keys(this.dayMap)
                    .filter(key => meetingTime[key])
                    .map(key => this.dayMap[key]);
            }
        }
    }

}
